src/java.base/share/classes/sun/security/util/DerInputStream.java

```diff
@@ -661,4 +684,118 @@ static int getDefiniteLength(InputStream in) throws IOException {
      * empty.
      */
     public int available() { return buffer.available(); }
+
+    /**
+     * Ensure there is no more data. This can be called when the last
+     * expected field is parsed and we need to make sure no unread is left.
+     */
+    public void atEnd() throws IOException {
+        if (available() != 0) {
+            throw new IOException("Extra unused bytes");
+        }
+    }
+
+    /**
+     * Detect if the tag of the next DerValue in the stream matches the rule.
+     *
+     * Attention: tag is an integer casted from a byte. i.e. could be negative.
+     *
+     * @param rule the rule to check for the tag.
+     * @return true if matches, false otherwise or stream is at end.
+     * @throws IOException if an I/O error happens
+     */
+    public boolean seeOptional(Predicate<Integer> rule) throws IOException {
```

Comment:
rule may be too generic? Probably something like tagChecker?

Reply:
There are other methods based on this one.